detectron2/utils/visualizer.py

```diff
@@ -1228,3 +1233,311 @@ def get_output(self):
             to the image.
         """
         return self.output
+
+
+class WandbVisualizer(DatasetEvaluator):
+
+    def __init__(self, dataset_name, size = -1) -> None:
+        super().__init__()
+        if wandb is None:
+            raise Exception("WandbVisualizer requires wandb. please install using `pip install wandb`")
+
+        self.dataset_name = dataset_name
+        self.size = size
+        self._run = None
+        # Table logging utils
+        self._evalset_table = None
+        self._evalset_table_rows = [] # reset after log operation
+        self._evalset_table_ref = None # eval artifact refrence for deduping. Don't reset 
+        self._row_idx = 0
+        self._map_table_row_file_name = {}
+        self._table_thing_classes = None
+        self._table_stuff_classes = None
+        
+        # parsed metadata
+        self.thing_class_names = []
+        self.thing_index_to_class = {}
+        self.stuff_class_names = []
+        self.stuff_index_to_class = {}
+
+
+    def process(self, inputs, outputs):
+        if self.size > -1 and len(self._evalset_table_rows) >= self.size // (comm.get_world_size()):
+            return
+        parsed_output = self._parse_prediction(outputs[0])
+        parsed_output["file_name"] = inputs[0].get("file_name")
+        table_row = self._plot_table_row(parsed_output)
+        self._evalset_table_rows.append(table_row)
+        if self._evalset_table is None:
+            self._evalset_table = self._build_evalset_table(parsed_output.keys())
+
+
+    def evaluate(self):
+        comm.synchronize()
+        table_rows = comm.gather(self._evalset_table_rows)
+        if comm.is_main_process():
+            if self._run is None:
+                self._run = wandb.init(project=self.dataset_name) if not wandb.run else wandb.run
```

Comment:
where will it log to when `not wandb.run` ?
If there isn't a reasonable place to log to, we shouldn't just create a random directory or some random project.

We could in `__init__` check if `wandb.run` is set, and raise an error. That means users are responsible for calling `wandb.init` or `WandbWriter` before using this evaluator.

An alternative is to let this class accept kwargs as well like `WandbWriter` - but then I think the semantics becomes unclear when both classes are used: one of them (not sure which) will actually do init and the other will completely ignore the kwargs.

Reply:
How about instead of all kwargs we let the `WandbVisualizer` take an argument `output_dir` like `COCOEvaluator` class and use that as output folder if wandb isn't explicitly initialized by the user?

Reply:
That also makes sense, but are you proposing to do this for both `WandbVisualizer` and `WandbWriter`, or just one? Consistency of usage is important and I can't see why one class accepts `**kwargs` for `wandb.init`, while another accepts `output_dir`.

Reply:
Yeah, fair point. 
So, I guess the only option that remains is to make sure a wandb run is initialized manually when Visualizer is used. We cannot rely on `WandbWriter`'s initialization because `build_writers` is called after evaluators are initialized.

Reply:
Yeah that's fair enough.
I changed the design so the visualizer won't call wandb.init. I relies on user either calling `wandb.init` manually or explicitly passing `wandbWriter`
```
 if comm.is_main_process():
       if self.wandb.run is None:
           raise Exception("wandb run is not initialized. Either call wandb.init(...) explicitly or set WandbWriter()")
       self._run = self.wandb.run
```

Reply:
@ppwwyyxx in case you missed ^^^